5 - Shor's Algorithm on logical cat qubits using the Qrisp framework.ipynb

```diff
@@ -0,0 +1,824 @@
+    "        )\n",
+    "    if r % 2:\n",
+    "        raise Exception(\n",
+    "            \"Found an odd period for the given base. Please retry with another value of a.\"\n",
+    "        )\n",
+    "    p = np.gcd(a ** (r // 2) + 1, N)\n",
+    "    q = int(N / p)\n",
+    "    if p > q:\n",
+    "        return (q, p)\n",
+    "    return (p, q)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0dee107e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "a = 2\n",
+    "N = 15\n",
+    "# WARNING: The code below takes approximatively 10 minutes to run\n",
+    "quantum_shor(a, N)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "91be5b03",
+   "metadata": {},
+   "source": [
+    "Notice that the parameter `a` is provided manually, instead of it being drawn randomly as in the classical case: this is merely due to optimisation reasons, as many runs of the algorithm are needed in order to find the proper `a`. This could take a long time to achieve!\n",
+    "\n",
+    "The remarkable aspect of the Qrisp framework is the compactness and simplicity of implementing a fairly complex algorithm, such as Shor's, in a few lines of code as well as the relatively high level of abstraction with respect to gate-level. This task is much more complex to achieve in Qiskit, for example. You can find either a [specific tutorial to factor 15 with an hard-coded circuit](https://github.com/qiskit-community/qiskit-community-tutorials/blob/master/algorithms/shor_algorithm.ipynb); either a [general implementation](https://github.com/Qiskit/qiskit/blob/9c8eb06984c91156eac078f3d2d73b4cf68370b4/qiskit/algorithms/factorizers/shor.py), which has become deprecated."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2b098bd5",
+   "metadata": {},
+   "source": [
+    "### But why is it so slow?\n",
+    "It takes of the order of minutes to factor $15$ with Shor's quantum algorithm.\n",
+    "\n",
+    "We first assume to run a noiseless classical simulation of a quantum algorithm. [Gottesman-Knill theorem](https://en.wikipedia.org/wiki/Gottesman%E2%80%93Knill_theorem) states that circuits composed of Clifford gates can be perfectly simulated in polynomial time on a probabilistic classical computer. Shor's algorithm, however, uses non-Clifford gates and, as a result, the simulation is much longer than in the classical case, in particular for small instances.\n",
+    "On top of that, when adding a noise model, the number of internal operations performed by the Qiskit's backend increases noticeably the computational cost, partially inficiating sparsity assumption of matrix operations and storage.\n",
+    "\n",
+    "In any case, Shor's prime factoring algorithm remains to this day the simplest algorithm to provide a theoretical super-polynomial speedup (if executed on an actual quantum computer) and actually observing it a work could give great insight into the realm of quantum computation.\n"
```

Comment:
"actually observing it work" you can remove the "a"

Reply:
I actually meant "at work" as "in action"